### PR TITLE
[Merged by Bors] - feat: No wrap-around principle for 3APs

### PR DESCRIPTION
**Pull request by YaelDillies**
Prove that `{0, ..., k - 1} ⊆ ℤ/(n + 1)ℤ` and `{0, ..., k - 1} ⊆ ℕ` are `m`-Freiman isomorphic if `m * k ≤ n`

---

- [x] depends on: #12542
- [x] depends on: #12546

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/CharP/Basic.lean [hunk 1]
```diff
@@ -91,7 +91,8 @@ theorem exists_add_pow_prime_eq (hp : p.Prime) (x y : R) :
 
 end CommSemiring
 
-variable (R)
+section AddMonoidWithOne
+variable (R) [AddMonoidWithOne R] (p : ℕ)
 
 /-- The generator of the kernel of the unique homomorphism ℕ → R for a semiring R.
```
Comment (TFBloom) [inline Mathlib/Algebra/CharP/Basic.lean:98]: The changes to CharP/Basic.lean seem more general and basic than stuff about 3APs, do they not belong in a separate PR?

**Review by TFBloom — CHANGES_REQUESTED** [review]

File: Mathlib/Algebra/CharP/Basic.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Algebra/CharP/Basic.lean:98]: See the PR description, in particular the mention of #12542!

File: Mathlib/Combinatorics/Additive/FreimanHom.lean [hunk 2]
```diff
@@ -289,6 +291,50 @@ lemma IsMulFreimanHom.inv (hf : IsMulFreimanHom n A B f) : IsMulFreimanHom n A B
...
+
+private lemma aux (hm : m ≠ 0) (hkmn : m * k ≤ n) : k < n.succ :=
+  Nat.lt_succ_iff.2 $ le_trans (Nat.le_mul_of_pos_left _ hm.bot_lt) hkmn
+
+/-- **No wrap-around principle**.
+
+The first `k + 1` elements of `Fin (n + 1)` are `m`-Freiman isomorphic to the first `k + 1` elements
+of `ℕ` assuming there is no wrap-around. -/
+lemma isAddFreimanIso_Iic (hm : m ≠ 0) (hkmn : m * k ≤ n) :
+    IsAddFreimanIso m (Iic (k : Fin n.succ)) (Iic k) val where
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Additive/FreimanHom.lean:305]: `n.succ` is no longer generated by `induction`, and it's also not really SNF. Is there a specific reason why you prefer it over `n+1`? Also elsewhere.

**Review by TFBloom — APPROVED** [review]
Everything looks good on the maths side!

File: Mathlib/Combinatorics/Additive/FreimanHom.lean [hunk 3]
```diff
@@ -289,6 +291,50 @@ lemma IsMulFreimanHom.inv (hf : IsMulFreimanHom n A B f) : IsMulFreimanHom n A B
...
+    exact ⟨congr_arg _, CharP.natCast_injOn_Iio _ (n + 1) (this hsA hs) (this htA ht)⟩
+
+/-- **No wrap-around principle**.
+
+The first `k` elements of `Fin (n + 1)` are `m`-Freiman isomorphic to the first `k` elements of `ℕ`
+assuming there is no wrap-around. -/
+lemma isAddFreimanIso_Iio (hm : m ≠ 0) (hkmn : m * k ≤ n) :
+    IsAddFreimanIso m (Iio (k : Fin (n + 1))) (Iio k) val := by
+  obtain _ | k := k
+  · simp [← bot_eq_zero]; simp [← _root_.bot_eq_zero, -bot_eq_zero']
```
Comment (Ruben-VandeVelde) [inline Mathlib/Combinatorics/Additive/FreimanHom.lean:327]: This looks kinda brittle. Would a simp only or simp_rw be clearer?

Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/FreimanHom.lean:327]: No, it's really not brittle. The issue is that simp in Lean 4 really dumbly does not remove "forward `foo`" when you ask it to use "backward `foo`": If `foo` is a simp lemma then `simp [← foo]` will invariably loop (unless you're lucky and it closes the goal before hitting the loop.

Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/FreimanHom.lean:327]: https://github.com/leanprover/lean4/issues/4290

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**